### PR TITLE
[Reviewer: Graeme] Send to multiple hosts

### PR DESCRIPTION
**Pull request by rkd-msw**
This round-robins traffic between multiple IP addresses, which may be useful for scale testing this sprint.

### CODE REVIEW

File: src/call.cpp [hunk 1]
```diff
@@ -728,7 +727,25 @@ bool call::connect_socket_if_needed()
             ERROR_NO("Unable to bind UDP socket");
         }
     } else { /* TCP, SCTP or TLS. */
-        struct sockaddr_storage *L_dest = &remote_sockaddr;
+        /* Pick a random remote IP address from the list. Doing this here
+         * (just before opening a new connection) means that we'll round-robin our
```
Comment (graemerobertson) [inline src/call.cpp]: Doesn't round-robin mean something different to random, which is what we're actually doing?

File: src/call.cpp [hunk 2]
```diff
@@ -728,7 +727,25 @@ bool call::connect_socket_if_needed()
...
+         * connections across all IP addresses to which the host resolves. */
+
+        size_t index = rand() % remote_ips.size();
+        strcpy(remote_ip, remote_ips[index].c_str());
+
+        struct sockaddr_storage L_dest;
+        if (gai_getsockaddr(&L_dest, remote_ip, remote_port,
+                    AI_PASSIVE, AF_UNSPEC) != 0) {
+            ERROR("Could not get socket for IP address '%s'.\n",
+                    remote_ip);
```
Comment (graemerobertson) [inline src/call.cpp]: Does `ERROR` exit here?

File: src/logger.cpp [hunk 3]
```diff
@@ -351,13 +351,14 @@ void print_stats_in_file(FILE * f)
...
         }
+
         fprintf(f, "   Port   Total-time  Total-calls  Remote-host" SIPP_ENDL
                 "%19s   %-5d %6lu.%02lu s     %8llu  %s:%d(%s)" SIPP_ENDL SIPP_ENDL,
                 temp_str,
                 local_port,
                 clock_tick / 1000, (clock_tick % 1000) / 10,
                 total_calls,
-                remote_ip,
+                remote_ips.empty() ? "" : remote_ips[0].c_str(),
```
Comment (graemerobertson) [inline src/logger.cpp]: This looks misleading

File: src/socket.cpp [hunk 4]
```diff
@@ -69,6 +69,35 @@ int pending_messages = 0;
 
 map<string, SIPpSocket *>     map_perip_fd;
 
+static int resolve_ips(std::string host, std::vector<std::string>& out)
+{
+    const struct addrinfo hints = {AI_PASSIVE, AF_UNSPEC,};
+    struct addrinfo* res;
+    int error = getaddrinfo(host.c_str(), NULL, &hints, &res);
+    // 40 characters is enough to store an IPv4 or IPv6 address.
+    char host[40];
```
Comment (graemerobertson) [inline src/socket.cpp]: I literally don't understand this - have you just somehow redefined what the variable `host` is?!

File: src/socket.cpp [hunk 5]
```diff
@@ -2506,19 +2537,15 @@ int open_connections()
...
             }
 
-            get_inet_address(&remote_sockaddr, remote_ip, sizeof(remote_ip));
-            if (remote_sockaddr.ss_family == AF_INET) {
-                strcpy(remote_ip_escaped, remote_ip);
-            } else {
-                sprintf(remote_ip_escaped, "[%s]", remote_ip);
-            }
+            // Use the first remote IP by default.
+            gai_getsockaddr(&remote_sockaddr, remote_ips[0].c_str(), remote_port,
```
Comment (graemerobertson) [inline src/socket.cpp]: What's this for? What is "default" used for?

File: src/call.cpp (hunk 2 above)
Comment (graemerobertson) [inline src/call.cpp]: Ooh, also alignment!

Comment (rkd-msw) [inline src/call.cpp]: Yes, it does.

File: src/socket.cpp (hunk 4 above)
Comment (rkd-msw) [inline src/socket.cpp]: Yeah, actually, pretty sure this didn't compile and I'd made but not pushed a commit to fix it.

**Pull request merged**